File: leetcode/Python/copy-list-with-random-pointer.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        v, cur, um, idx = list(), head, dict(), 0
        while cur:
            um[cur], t = idx, Node(cur.val)
            if len(v) > 0:
                v[-1].next = t
            v.append(t)
            cur, idx = cur.next, idx+1
        cur, idx = head, 0
        while cur:
            if cur.random is not None:
                v[idx].random = v[um[cur.random]]
            cur, idx = cur.next, idx + 1
        return v[0] if len(v)>0 else None

File: leetcode/C++/copy-list-with-random-pointer.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/

class Solution {
public:
    Node* copyRandomList(Node* head) {
        unordered_map<Node*, int> um;
        vector<Node*> v;
        int idx=0;
        for(Node* cur=head, *t=nullptr; cur; cur=cur->next) {
            um[cur]=idx++;
            t = new Node(cur->val);
            if(!v.empty()) v.back()->next = t;
            v.push_back(t);
        }
        idx=0;
        for(Node* cur=head; cur; cur=cur->next,++idx) if(cur->random) v[idx]->random=v[um[cur->random]];
        return (!v.empty())? v[0] : nullptr;
    }
};